### DISCUSSION TITLE
How to pkg theia browser to an excutable application？

### DISCUSSION THREAD

**Discussion by SunnyFitz** [eclipse-theia/theia#9456] [Q&A]
I want to use node module “pkg” to package theia browser to an excutable application. But it Cannot find module while compile . Is there some documents to guide us to pkg theia project ? Thanks.

**Comment by vince-fugnitto** [comment]
@SunnyFitz thank you for the discussion, there is no documentation on how to do so, we are currently in the process of packaging a theia-based application ourselves with `pkg` for `eclipse theia blueprint`: https://github.com/eclipse-theia/theia-blueprint/issues/74.

We have identified necessary updates  to the framework to allow packaging the application with `pkg`, as well as updates to `pkg` itself (which will allow language-servers to be correctly spawned).

**Reply by ballcoach12** [reply]
@vince-fugnitto, since I still consider myself new to Theia, I'd like to ask how closely what you envision matches the workflow that I need to support. My apologies if my comment is slightly off topic; I'll be happy to move my question into a new discussion if that's more appropriate.

1. We want to build custom Theia editors for various DSLs that we support. The DSLs map to products that my company sells.

2. Just like in the existing Theia Blueprint implementation, we'd like to specify extensions that should be built into the final editor

3. As a minimum, we'd like to build a NodeJS app that runs Theia in a browser and can be packaged in a Dockerfile

4. As an option, we'd like to build a desktop Electron app just like Blueprint does now.

5. Ultimately, we'd like to create a Yeoman script that generates the entire project infrastructure based on input from the user.

It sounds like the idea you guys are pursuing to use `pkg` is somewhat similar, without the automation that I describe in item 5. 

Can you comment on how close my concept is to what you are currently pursuing?

Thanks!

**Reply by vince-fugnitto** [reply]
@ballcoach12 no problem, I'll try and answer as best I can!

> 1. We want to build custom Theia editors for various DSLs that we support. The DSLs map to products that my company sells.

Sounds good, the following info might be interesting to you: 
- https://www.typefox.io/blog/domain-specific-languages-in-theia-and-vs-code

> 2. Just like in the existing Theia Blueprint implementation, we'd like to specify extensions that should be built into the final editor

:+1: 

> 3. As a minimum, we'd like to build a NodeJS app that runs Theia in a browser and can be packaged in a Dockerfile

The ability to package the browser-app as an executable and distribute it in a dockerfile does not yet exist, it is what https://github.com/eclipse-theia/theia-blueprint/issues/74 is attempting to make an example for. For the moment, we can still have a theia application in a dockerfile, but it is built as a layer in the image. We have some soon-to-be deprecated examples in theia-apps:
- https://github.com/theia-ide/theia-apps/tree/master/theia-docker

> 4. As an option, we'd like to build a desktop Electron app just like Blueprint does now.

This is possible today, you can have both a browser and electron based version of your application with the same sources.
In this repo, we use both versions of the application from the same sources for test purposes:
- https://github.com/eclipse-theia/theia/tree/master/examples

> 5. Ultimately, we'd like to create a Yeoman script that generates the entire project infrastructure based on input from the user.

I do not believe this point exists today, we only for example have a generator to help adopters quickly bootstrap their own custom theia extensions. I'm not sure what you'd like your users to input and what you'd want the outcome to be, do you mind clarifying? It might be something you can implement yourselves if needed as well.

> Can you comment on how close my concept is to what you are currently pursuing?

The concept (https://github.com/eclipse-theia/theia-blueprint/issues/74) is to be able to package the browser based application as an executable which can then be more easily consumed in a docker image. If you have further questions about that initiative itself, please feel free to comment on the issue itself where the authors might be able to give you a clearer picture :+1:

**Reply by ballcoach12** [reply]
Thanks for the quick reply, @vince-fugnitto. I should have clarified that we'd plan to take on item 5 ourselves, since it would be tailored towards our internal development teams. 

I am currently evaluating [https://github.com/eclipse-theia/theia-blueprint/issues/74](https://github.com/eclipse-theia/theia-blueprint/issues/74) as a possible starting point for our implementation, and it looks promising. Since you were a reviewer, I took a chance and asked questions here.  If I have further questions about that issue, then I'll post them on the theia-blueprint project.
